src/common/selftest/selftest_esp_update.cpp

```diff
@@ -0,0 +1,344 @@
+        case esp_upload_action::Initial:
+            phase = from_menu ? PhasesSelftest::ESP_ask_from_menu : PhasesSelftest::ESP_ask_auto;
+            progress_state = esp_upload_action::Initial_wait_user;
+            break;
+        case esp_upload_action::Initial_wait_user:
+            if (continue_pressed) {
+                espif_flash_initialize();
+                struct stat fs;
+                for (esp_entry *chunk = firmware_set.begin();
+                     chunk != firmware_set.end(); ++chunk) {
+                    if (stat(chunk->filename, &fs) == 0) {
+                        chunk->size = fs.st_size;
+                    }
+                }
+                progress_state = esp_upload_action::Connect;
+                phase = PhasesSelftest::ESP_upload;
+            }
+            break;
+        case esp_upload_action::Connect: {
+            esp_loader_connect_args_t config = ESP_LOADER_CONNECT_DEFAULT();
+            if (ESP_LOADER_SUCCESS == esp_loader_connect(&config)) {
+                log_info(Network, "ESP boot connect OK");
+                progress_state = esp_upload_action::Start_flash;
+            } else {
+                log_debug(Network, "ESP boot failedK");
+                progress_state = esp_upload_action::ESP_error;
+            }
+            break;
+        }
+        case esp_upload_action::Start_flash:
+            if ((opened_file = fopen(current_file->filename, "rb")) == nullptr) {
+                log_error(Network, "ESP flash: Unable to open file %s", current_file->filename);
+                progress_state = esp_upload_action::USB_error;
+                break;
+            } else {
+                log_info(Network, "ESP Start flash %s", current_file->filename);
+                updateProgress();
+            }
+
+            if (esp_loader_flash_start(
+                    current_file->address,
+                    current_file->size,
+                    buffer_length)
+                != ESP_LOADER_SUCCESS) {
+                log_error(Network, "ESP flash: Unable to start flash on address %0xld", current_file->address);
+                progress_state = esp_upload_action::ESP_error;
+                fclose(opened_file);
+                break;
+            } else {
+                progress_state = esp_upload_action::Write_data;
+                readCount = 0;
+                ++current_file_no;
+            }
+            break;
+        case esp_upload_action::Write_data: {
+            size_t readBytes = 0;
+            uint8_t buffer[buffer_length];
+
+            readBytes = fread(buffer, 1, sizeof(buffer), opened_file);
+            readCount += readBytes;
+            log_debug(Network, "ESP read data %ld", readCount);
+            if (ferror(opened_file)) {
+                log_error(Network, "ESP flash: Unable to read file %s", current_file->filename);
+                readBytes = 0;
+                progress_state = esp_upload_action::USB_error;
+                break;
```

Comment:
Where do we close `opened_file` in such a case? I'd recommend using some FILE RAII to avoid resource leak

Reply:
fixed